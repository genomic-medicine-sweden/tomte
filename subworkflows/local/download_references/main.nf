//
// Download reference files
//

include { GENCODE_DOWNLOAD as FASTA_DOWNLOAD } from '../../../modules/local/download/gencode_download'
include { GENCODE_DOWNLOAD as GTF_DOWNLOAD   } from '../../../modules/local/download/gencode_download'
include { WGET_DOWNLOAD                      } from '../../../modules/local/download/wget_download'
include { VEP_GNOMAD_DOWNLOAD                } from '../../../modules/local/download/vep_gnomad_download'
include { BUILD_VEP_CACHE                    } from '../../../modules/local/build_vep_cache'

workflow DOWNLOAD_REFERENCES {
    take:
        ch_genome                        // channel: [mandatory]  val(genome)
        ch_gencode_annotation_version    // channel: [mandatory]  val(gencode_annotation_version)
        ch_vep_refs_download_unprocessed // channel: [optional]   val(path_to_csv)
        ch_vep_cache_version             // channel: [optional]   val(vep_cache_version)

    main:
        ch_versions = Channel.empty()

        // Download fasta if not provided
        ch_downloaded_fasta = Channel.empty()
        if ( !params.fasta ) {
            FASTA_DOWNLOAD(ch_genome, ch_gencode_annotation_version, "fasta")
            ch_downloaded_fasta = ch_downloaded_fasta.mix( FASTA_DOWNLOAD.out.fasta ).collect()
            ch_versions = ch_versions.mix(FASTA_DOWNLOAD.out.versions)
        }

        // Download gtf if not provided
        ch_downloaded_gtf = Channel.empty()
        if ( !params.gtf ) {
            GTF_DOWNLOAD(ch_genome, ch_gencode_annotation_version, "gtf")
            ch_downloaded_gtf = ch_downloaded_gtf.mix( GTF_DOWNLOAD.out.gtf ).collect()
            ch_versions = ch_versions.mix(GTF_DOWNLOAD.out.versions)
        }

        // Read and store paths in vep_refs_download_unprocessed
        // Download files
        if ( !params.skip_download_vep && !params.vep_cache ) { 
            ch_vep_refs_download_unprocessed.splitCsv(header: true)
                .map { row -> return tuple(row.name, row.path_for_wget) }
                .set { ch_vep_refs_download }
            WGET_DOWNLOAD(ch_vep_refs_download.filter{ it != null })
            ch_versions = ch_versions.mix(WGET_DOWNLOAD.out.versions)
        }

        if ( !params.skip_download_gnomad ) {
            VEP_GNOMAD_DOWNLOAD(ch_genome, ch_vep_cache_version)
            ch_versions = ch_versions.mix(VEP_GNOMAD_DOWNLOAD.out.versions)
        }

        ch_built_vep_cache = Channel.empty()
        ch_built_vep_plugin_file = Channel.empty()
        if ( !params.skip_download_vep && !params.vep_cache ){
            BUILD_VEP_CACHE(WGET_DOWNLOAD.out.downloaded_file.collect(), VEP_GNOMAD_DOWNLOAD.out.gnomad_vcf_tbi.flatten().collect())
            ch_built_vep_cache = ch_built_vep_cache.mix( BUILD_VEP_CACHE.out.vep_cache )
            ch_built_vep_plugin_file = ch_built_vep_plugin_file.mix( BUILD_VEP_CACHE.out.vep_plugin_file )
            ch_versions = ch_versions.mix(BUILD_VEP_CACHE.out.versions)
        }


    emit:
        fasta      = ch_downloaded_fasta      // channel: [ path(fasta) ]
        gtf        = ch_downloaded_gtf        // channel: [ path(gtf) ]
        vep_cache  = ch_built_vep_cache       // channel: [ path(vep_cache) ]
        vep_plugin = ch_built_vep_plugin_file // channel: [ path(vep_plugin) ]
        versions   = ch_versions              // channel: [ path(versions.yml) ]
}
