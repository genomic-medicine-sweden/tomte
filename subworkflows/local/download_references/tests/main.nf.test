nextflow_workflow {

    name "Test Workflow DOWNLOAD_REFERENCES"
    script "../main.nf"
    workflow "DOWNLOAD_REFERENCES"
    tag "subworkflow"
    tag "download_references"

    test("Should run without failures") {
        options "-stub"

        when {
            params {
                skip_download_gnomad = false
                skip_download_vep    = false
                fasta                = []
                gtf                  = []
                vep_cache            = []
            }
            workflow {
                """
                input[0] = 'GRCh37'
                input[1] = Channel.of(46)
                input[2] = Channel.of([file("${projectDir}/test_data/vep_to_download_37.csv", checkIfExists: true)])
                input[3] = Channel.of(112)
                input[4] = true
                input[5] = true
                input[6] = true
                input[7] = true
                input[8] = true
                """
            }
        }

        then {
            assertAll(
                { assert workflow.success },
                { assert snapshot(
                    workflow.out.fasta,
                    workflow.out.vep_cache,
                    workflow.out.vep_plugin,
                    file(workflow.out.high_q_vcf[0]).name,
                    file(workflow.out.high_q_vcf[1]).name,
                    workflow.out.versions,
                ).match()}
            )
        }
    }
}
