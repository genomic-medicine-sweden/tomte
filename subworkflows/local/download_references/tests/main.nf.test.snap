{
    "Should run without failures": {
        "content": [
            {
                "0": [
                    
                ],
                "1": [
                    
                ],
                "2": [
                    [
                        
                    ]
                ],
                "3": [
                    "vep_plugin_files.csv:md5,d41d8cd98f00b204e9800998ecf8427e"
                ],
                "4": [
                    "versions.yml:md5,7e36dbe96e8aafd3ba9bf1bd0e787a09",
                    "versions.yml:md5,7e36dbe96e8aafd3ba9bf1bd0e787a09",
                    "versions.yml:md5,7e36dbe96e8aafd3ba9bf1bd0e787a09",
                    "versions.yml:md5,e8e8c4e637cdc23a54d8fb5e8958c4bd"
                ],
                "fasta": [
                    
                ],
                "gtf": [
                    
                ],
                "vep_cache": [
                    [
                        
                    ]
                ],
                "vep_plugin": [
                    "vep_plugin_files.csv:md5,d41d8cd98f00b204e9800998ecf8427e"
                ],
                "versions": [
                    "versions.yml:md5,7e36dbe96e8aafd3ba9bf1bd0e787a09",
                    "versions.yml:md5,7e36dbe96e8aafd3ba9bf1bd0e787a09",
                    "versions.yml:md5,7e36dbe96e8aafd3ba9bf1bd0e787a09",
                    "versions.yml:md5,e8e8c4e637cdc23a54d8fb5e8958c4bd"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "24.10.5"
        },
        "timestamp": "2025-04-14T12:02:19.39737"
    }
}